### PR TITLE
feat: simplify the omnichain template

### PR DESCRIPTION
**Pull request by fadeev**
Before:

```solidity
// SPDX-License-Identifier: MIT
pragma solidity 0.8.7;
 
import "@zetachain/protocol-contracts/contracts/zevm/SystemContract.sol";
import "@zetachain/protocol-contracts/contracts/zevm/interfaces/zContract.sol";
import "@zetachain/toolkit/contracts/OnlySystem.sol";
 
contract MyContract is zContract, OnlySystem {
    SystemContract public systemContract;
 
    constructor(address systemContractAddress) {
        systemContract = SystemContract(systemContractAddress);
    }
 
    function onCrossChainCall(
        zContext calldata context,
        address zrc20,
        uint256 amount,
        bytes calldata message
    ) external virtual override onlySystem(systemContract) {
        // TODO: implement the logic
    }
}
```

After:

```solidity
// SPDX-License-Identifier: MIT
pragma solidity 0.8.7;

import "@zetachain/toolkit/contracts/ZetaChainApp.sol";

contract Swap is ZetaChainApp {
    constructor() ZetaChainApp() {}

    function onCrossChainCall(
        zContext calldata context,
        address zrc20,
        uint256 amount,
        bytes calldata message
    ) external override onlySystem {
        // TODO: implement the logic
    }
}
```

### CODE REVIEW

File: contracts/ZetaChain.sol [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+pragma solidity 0.8.7;
+
+import "./OnlySystem.sol";
+import "./BytesHelperLib.sol";
+import "./SwapHelperLib.sol";
+import "@zetachain/protocol-contracts/contracts/zevm/SystemContract.sol";
+import "@zetachain/protocol-contracts/contracts/zevm/interfaces/IWZETA.sol";
+
+abstract contract ZetaChain is zContract, OnlySystem {
+    SystemContract public systemContract;
```
Comment (andresaiello) [inline contracts/ZetaChain.sol:11]: you inherit from OnlySystem the system contract address; looks redundant to have this one also.

Comment (fadeev) [reply contracts/ZetaChain.sol:11]: I think we need this variable to able to use it from other functions without having to pass it explicitly, for example:

https://github.com/zeta-chain/toolkit/pull/135

Comment (fadeev) [reply contracts/ZetaChain.sol:11]: `OnlySystem` only has system contract address, `systemContract` on the other hand is a reference to the system contract, which then may be used in other functions like so https://github.com/zeta-chain/toolkit/pull/135

Open to suggestions on how to streamline this.

Comment (fadeev) [reply contracts/ZetaChain.sol:11]: I suppose we can do

```solidity
abstract contract ZetaChain is zContract, OnlySystem {
    SystemContract public systemContract;
    mapping(uint256 => address) private systemContracts;

    constructor() OnlySystem(_initializeSystemContracts()) {
        systemContract = SystemContract(systemContractAddress);
    }

    function _initializeSystemContracts() private returns (address) {
        systemContracts[7000] = 0x91d18e54DAf4F677cB28167158d6dd21F6aB3921;
        systemContracts[7001] = 0xEdf1c3275d13489aCdC6cD6eD246E72458B8795B;
        return systemContracts[block.chainid];
    }
}
```

Comment (fadeev) [reply contracts/ZetaChain.sol:11]: The example of why I think we need a reference to the system contract is:

```solidity
abstract contract ZetaChain is zContract, OnlySystem {
    SystemContract public systemContract;
    mapping(uint256 => address) private systemContracts;

    constructor() OnlySystem(_initializeSystemContracts()) {
        systemContract = SystemContract(systemContractAddress);
    }

    function _initializeSystemContracts() private returns (address) {
        systemContracts[7000] = 0x91d18e54DAf4F677cB28167158d6dd21F6aB3921;
        systemContracts[7001] = 0xEdf1c3275d13489aCdC6cD6eD246E72458B8795B;
        return systemContracts[block.chainid];
    }

    function doSomething() public {
        // systemContract.uniswapv2Router02Address()
    }
}
```

Comment (andresaiello) [reply contracts/ZetaChain.sol:11]: then let's reference the contract in OnlySystem. I don't like to use two memory slot for exactly the same data.

Comment (fadeev) [reply contracts/ZetaChain.sol]: Moved all the system contract logic to `OnlySystem`: d4ecf31

**Review by andresaiello — APPROVED** [review]

**Pull request closed without merging**